### PR TITLE
Reland: Enable `dim=None` for `torch.sum`

### LINKED ISSUE #29137: torch.sum(tensor, dim=()) is different from np.sum(arr, axis=())
## 🐛 Bug

`torch.sum(tensor, dim=())` performs a full reduce, while `np.sum(arr, axis=())` performs no reduce.

## To Reproduce

```
import torch
import numpy as np
arr = np.array([1, 2, 3])
tensor = torch.from_numpy(arr)

tensor.sum(())  # gives tensor(6)
arr.sum(())  # gives np.array([1, 2, 3])
```
## Expected behavior

These should probably be the same. Also, I don't think it makes sense for `tensor.sum(dim=())` to do a full reduce. This special casing caused the bug at https://github.com/pytorch/pytorch/issues/28993 and I suspect the gradient for things like torch.var may be incorrect right now.

## Environment

Pytorch master

## Additional context

Some people who have been involved in multidim support who may be interested: @t-vi, @SsnL, @umanwizard @gchanan. Please let me know your opinions on this subject.

cc @ezyang @gchanan @zou3519 @jerryzh168 @SsnL

### PR DESCRIPTION
**Pull request by kurtamohler**
Part of #29137

Reland of #75845

### CODE REVIEW

**Comment by kurtamohler** [comment]
For some reason, the `macos-10-15-py3-arm64` (one of the jobs that failed after the initial merge) isn't being run. It did run the first time when I created this PR with the `ciflow/trunk` and `ciflow/periodic` labels. But after updating the PR, it doesn't get included anymore

**Comment by albanD** [comment]
It is there now :)

**Comment by ngimel** [comment]
cc @zou3519, this will also break functorch

**Comment by zou3519** [comment]
I can handle landing this

**Comment by zou3519** [comment]
While I figure out what the functorch-side changes should be, is there anyone else we need to coordinate a land with? pytorch/xla?

**Comment by zou3519** [comment]
I checked running this PR against functorch and it looks like it doesn't break functorch (despite the schema changes). So we can land it via the regular GHF route, whenever we figure out the coordination with XLA

**Comment by ngimel** [comment]
That's weird, last time I tried to build functorch in the brief period the PR was in trunk, I couldn't.

**Comment by kurtamohler** [comment]
I just tried building functorch with this change, and it seems to have worked. A few of the functorch tests fail, but the ones I saw don't appear to be related to `sum`

**Comment by zou3519** [comment]
This PR LGTM from a functorch perspective. I built both OSS functorch + this PR, and internal functorch with an imported version of the PR and they look fine

**Comment by kurtamohler** [comment]
The [win-vs2019-cuda11.6-py3 / test (default, 4, 5, windows.8xlarge.nvidia.gpu)](https://github.com/pytorch/pytorch/runs/7031617009?check_suite_focus=true#logs) job failed on `test_matmul_45724_cuda` from `test_linalg.py`, which doesn't use `sum` at all. I think it's probably a flaky test

**Comment by kurtamohler** [comment]
I think this PR is ready. Once it's approved, I can coordinate with XLA again

**Comment by zou3519** [comment]
AFAICT the only thing that changed here since the reland was the MPS changes, right? Who should review that?

**Comment by kurtamohler** [comment]
> AFAICT the only thing that changed here since the land was the MPS changes, right? Who should review that?

Yep, that's right. I'm not sure who

**Comment by zou3519** [comment]
cc @albanD ?

**Review by albanD — APPROVED** [review]
MPS side change LGTM
cc @kulinseth for visitbility

**Comment by kurtamohler** [comment]
@JackCaoG , this PR is ready. Can we coordinate again for the XLA PR?

**Comment by kurtamohler** [comment]
Sounds good, thanks!

**Review by kulinseth — APPROVED** [review]
MPS reduction changes look fine to me.

**Comment by JackCaoG** [comment]
@kurtamohler I think I have fixed the xla side head issue. Can you rebase this pr(to avoid false alarm) and I will rebase my xla pr and rerun all test. If everything is green we are good to go!

**Comment by kurtamohler** [comment]
@JackCaoG, PR has been rebased

**Comment by JackCaoG** [comment]
@kurtamohler xla pr is green, we can coordindate a reland tmr.
